core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -669,16 +669,824 @@ impl CheckedAdd for Fixed64 {
+				Zero::zero()
+			}
+		}
+
+		/// Number of limbs.
+		pub fn len(&self) -> usize { self.digits.len() }
+
+		/// A naive getter for limb at `index`. Note that the order is lsb -> msb.
+		///
+		/// #### Panics
+		///
+		/// This panics if index is out of range.
+		pub fn get(&self, index: usize) -> Single {
+			let len = self.digits.len();
+			self.digits[len - 1 - index]
+		}
+
+		/// A naive setter for limb at `index`. Note that the order is lsb -> msb.
+		///
+		/// #### Panics
+		///
+		/// This panics if index is out of range.
+		pub fn set(&mut self, index: usize, value: Single) {
+			let len = self.digits.len();
+			self.digits[len - 1 - index] = value;
+		}
+
+		/// returns the least significant limb of the number.
+		///
+		/// #### Panics
+		///
+		/// While the constructor of the type prevents this, this can panic if `self` has no digits.
+		pub fn lsb(&self) -> Single {
+			self.digits[self.len() - 1]
+		}
+
+		/// returns the most significant limb of the number.
+		///
+		/// #### Panics
+		///
+		/// While the constructor of the type prevents this, this can panic if `self` has no digits.
+		pub fn msb(&self) -> Single {
+			self.digits[0]
+		}
+
+		/// Strips zeros from the left side of `self`, if any.
+		pub fn lstrip(&mut self) {
+			// by definition, a big-int number should never have leading zero limbs. This function
+			// has the ability to cause this. There is nothing to do if the number already has 1
+			// limb only. call it a day and return.
+			if self.len().is_zero() { return; }
+			let mut index = 0;
+			for elem in self.digits.iter() {
+				if *elem != 0 { break } else { index += 1 }
+			}
+			if index > 0 {
+				self.digits = self.digits[index..].to_vec()
+			}
+		}
+
+		/// Zero-pad `self` from left to reach `size` limbs. Will not make any difference if `self`
+		/// is already bigger than `size` limbs.
+		pub fn lpad(&mut self, size: usize) {
+			let n = self.len();
+			if n >= size { return; }
+			let pad = size - n;
+			let mut new_digits = (0..pad).map(|_| 0).collect::<Vec<Single>>();
+			new_digits.extend(self.digits.iter());
+			self.digits = new_digits;
+		}
+
+		/// Extends either one of `self` or `other` to meet the size of the other one. Equivalent of
+		/// using `lpad()` of the smaller one with the size of the bigger one.
+		pub fn resize(&mut self, other: &mut Self) {
+			if self.len() >= other.len() {
+				other.lpad(self.len())
+			} else if other.len() > self.len() {
+				self.lpad(other.len())
+			}
+		}
+
+		/// Adds `self` with `other`. self and other do not have to have any particular size. Given
+		/// that the `n = max{size(self), size(other)}`, it will produce a number with `n + 1`
+		/// limbs.
+		///
+		/// This function does not strip the output and returns the original allocated `n + 1`
+		/// limbs. The caller may strip the output if desired.
+		///
+		/// Taken from "The Art of Computer Programming" by D.E. Knuth, vol 2, chapter 4.
+		pub fn add(mut self, mut other: Self) -> Self {
```

Comment:
Why does this consume `other` instead of taking `other: &Self`?

Reply:
I am generally more in favour of consuming when an arithmetic operation cannot fail. If it might fail, then it is more sensible to take a reference. 

But now seeing the big picture here, I suppose I could also take all the `rhs`s by reference.

Reply:
You don't need to resize if `get` returns a default of 0 for out-of-range indices.

Reply:
resize cannot change (it must change the caller somehow) and get should be a raw getter preferably, similar to other std types. We can have a `checked_get() -> Option<_>` and `unwrap_or` over it. 

I am quite doubtful still regarding accepting arguments as ref or owning. 
For the reason mentioned above, I prefer owning. But since this type can own a large vector of numbers moving stuff around is for sure non-sensical. Moreover, this type is still a _number_ and the std implementation of those takes everything by value (thou' this is not a strong argument as those all have `Copy`).

Hence, if we are to value speed over consistency, I don't see it fair to only make `other` a ref and still consume self. Both will be ref. I am in favour of implementing this.